### PR TITLE
Various Extensions

### PR DESCRIPTION
**Pull request by 5outh**
Various extensions for the new version

### CODE REVIEW

File: src/SellerLabs/Research/Entities/Item.php [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        );
+
+        $this->offerSummary = $this->parseOfferSummary(
+            $this->get('OfferSummary', [])
+        );
+
+        $this->offers = $this->parseOffers($this->get('Offers.Offer', []));
+    }
+
+    protected function parseItemLinks($itemLinks)
```
Comment (MarkVaughn) [inline src/SellerLabs/Research/Entities/Item.php]: docblox

File: src/SellerLabs/Research/Entities/Item.php [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    {
+        return  array_map(function ($item) {
+            return new ItemLink($item);
+        }, $itemLinks);
+    }
+
+    protected function parseItemAttributes($itemAttributes)
```
Comment (MarkVaughn) [inline src/SellerLabs/Research/Entities/Item.php]: ditto

File: src/SellerLabs/Research/Entities/Item.php [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    {
+        return new ItemAttributes($itemAttributes);
+    }
+
+    protected function parseOfferSummary($offerSummary)
```
Comment (MarkVaughn) [inline src/SellerLabs/Research/Entities/Item.php]: ditto

File: src/SellerLabs/Research/Entities/Item.php [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    {
+        return new OfferSummary($offerSummary);
+    }
+
+    protected function parseOffers(array $offers)
```
Comment (MarkVaughn) [inline src/SellerLabs/Research/Entities/Item.php]: ditto

File: src/SellerLabs/Research/Entities/ItemAttributes.php [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+ * @package SellerLabs\Research\Entities
+ *
+ * @author Benjamin Kovach <benjamin@roundsphere.com>
+ */
+class ItemAttributes extends BaseEntity
+{
+    /**
+     * @var array
+     */
+    public $attributes = [];
```
Comment (MarkVaughn) [inline src/SellerLabs/Research/Entities/ItemAttributes.php]: if you are accessing this with overloading this should be protected

File: src/SellerLabs/Research/Entities/ItemAttributes.php [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    /**
+     * Get an attribute or property (shorthand)
+     *
+     * @param string $name
+     *
+     * @return mixed
+     */
+    public function __get($name)
+    {
+        if ($name == 'attributes') {
```
Comment (MarkVaughn) [inline src/SellerLabs/Research/Entities/ItemAttributes.php]: this is unnecessary if attributes is public

File: src/SellerLabs/Research/Entities/RelationshipBag.php [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    protected $parents = [];
+
+    /**
+     * @var ProductRelationship[]
+     */
+    protected $children = [];
+
+    /**
+     * RelationshipBag constructor.
+     * @param $relationships
```
Comment (MarkVaughn) [inline src/SellerLabs/Research/Entities/RelationshipBag.php]: nit: newline & type

File: src/SellerLabs/Research/Entities/SearchProduct.php [hunk 8]
```diff
@@ -216,4 +221,28 @@ public function setWeight($weight)
...
+     */
+    public function setRankings(array $rankings)
+    {
+        $this->rankings = $rankings;
+    }
+
+    /**
+     * @param $relationships
+     */
+    public function setRelationships($relationships)
```
Comment (MarkVaughn) [inline src/SellerLabs/Research/Entities/SearchProduct.php]: setRelationship(RelationshipBag $relationship)  seing as getRelationships is returning that type

File: src/SellerLabs/Research/Entities/ItemAttributes.php [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+     * ItemAttributes constructor.
+     *
+     * @param array $raw
+     */
+    public function __construct(array $raw)
+    {
+        parent::__construct($raw);
+
+        foreach ($raw as $key => $val) {
+
```
Comment (etcinit) [inline src/SellerLabs/Research/Entities/ItemAttributes.php]: nit: Delete this empty line.

File: src/SellerLabs/Research/Factories/ItemLink.php [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+    public $description;
+
+    /**
+     * @var string
+     */
+    public $url;
+
+    /**
+     * ItemLink constructor.
+     * @param $link
```
Comment (etcinit) [inline src/SellerLabs/Research/Factories/ItemLink.php]: nit: New line before this.

File: src/SellerLabs/Research/ResearchClient.php [hunk 11]
```diff
@@ -237,4 +238,33 @@ public function getCategoryById($categoryId)
...
+
+    /**
+     * Hit the itemSearch endpoint of research-api with a given keyword
+     * phrase, page (default 1) and search index (default Blended)
+     *
+     * @param $keywords
+     * @param int $page
+     * @param string $searchIndex
+     *
+     * @return Responses\BaseResponse
```
Comment (etcinit) [inline src/SellerLabs/Research/ResearchClient.php]: suggestion: I tend to only use the base name of a class.

File: tests/SellerLabs/Research/Factories/ItemSearchResponseFactoryTest.php [hunk 12]
```diff
@@ -0,0 +1,18 @@
+<?php
+
+namespace Tests\SellerLabs\Research\Factories;
+
+use SellerLabs\Research\Factories\ItemSearchResponseFactory;
+
+class ItemSearchResponseFactoryTest extends ResponseFactoryTestCase
```
Comment (etcinit) [inline tests/SellerLabs/Research/Factories/ItemSearchResponseFactoryTest.php]: nit: Class comment is missing.

File: src/SellerLabs/Research/Entities/OfferListing.php [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+        $this->amountSaved = new Price(
+            $this->get('AmountSaved', [])
+        );
+        $this->percentageSaved = (int) $this->get('PercentageSaved', 0);
+        $this->availablity = $this->get('Availability');
+        $this->availabilityAttributes = new AvailabilityAttributes(
+            $this->get('AvailabilityAttributes')
+        );
+        $this->isEligibleForSuperSaverShipping = $this->get(
+            'IsEligibleForSuperSaverShipping'
```
Comment (tdouce) [inline src/SellerLabs/Research/Entities/OfferListing.php]: Does this return a `boolean`?

File: src/SellerLabs/Research/Entities/OfferListing.php [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+        );
+        $this->isEligibleForPrime = $this->get(
+            'IsEligibleForPrime'
+        );
```
Comment (tdouce) [inline src/SellerLabs/Research/Entities/OfferListing.php]: Does this return a `boolean`?

**Pull request merged**